src/UserSpaceInstrumentation/AccessTraceesMemory.cpp

```diff
@@ -25,44 +26,37 @@ using orbit_base::ReadFileToString;
 [[nodiscard]] ErrorMessageOr<std::vector<uint8_t>> ReadTraceesMemory(pid_t pid,
                                                                      uint64_t address_start,
                                                                      uint64_t length) {
-  // Round up length to next multiple of eight.
-  length = ((length + 7) / 8) * 8;
+  CHECK(length != 0);
+
+  OUTCOME_TRY(fd, orbit_base::OpenFileForReading(absl::StrFormat("/proc/%d/mem", pid)));
+
   std::vector<uint8_t> bytes(length);
-  for (size_t i = 0; i < length / 8; i++) {
-    const uint64_t data = ptrace(PTRACE_PEEKDATA, pid, address_start + i * 8, NULL);
-    if (errno) {
-      return ErrorMessage(
-          absl::StrFormat("Failed to PTRACE_PEEKDATA for pid %d with errno %d: \"%s\"", pid, errno,
-                          SafeStrerror(errno)));
-    }
-    std::memcpy(bytes.data() + (i * sizeof(uint64_t)), &data, 8);
+  OUTCOME_TRY(result, ReadFullyAtOffset(fd, bytes.data(), length, address_start));
```

Comment:
this error does not provide context. It will return something like "Permission denied." maybe add an if (result.has_error) { provide some context about what it was trying to do (read memory of a process with pid=%d) }

Reply:
My personal preference is not to use OUTCOME_TRY.

Reply:
I thought the same but started liking OUTCOME_TRY for keeping the code size under control. In many cases the error messages are good enough and it is reasonably clear what is going on. The caller still has the information that they called a "ReadTraceesMemory" and "Permission denied." is not the worst response there.

Reply:
the missing part here of course if that it was trying to read from /proc/pid/mem.